user-guide: multi-stage container (#291)
foundriesio/docs: source/user-guide/mult-stage-container/mult-stage-container.rst

.. _ug-mult-stage-container:

Use Multi-Stage Container Builds
================================

When you are developing for embedded devices optimization is one of the most 
important things to consider.

Multi-stage builds are useful to optimize your Dockerfiles as well as keep them 
easy to read and maintain.

You can find more information about it on: `Use multi-stage builds`_.

To summarize, multi-stage container builds allow you to develop in one stage and 

> copy just the necessary artifacts to the next stage. That been said, in the first 

Reviewer: ```suggestion
copy just the necessary artifacts to the next stage. That being said, in the first 
```

stage you can install dependencies for development like a toolchain and compile 
your application. In the final stage, you can keep the image nice and clean with 
just what is necessary for your application and copy the final artifacts from 
the first stage.

This section will present a Dockerfile implementing a Hello World application 

> written in C. In the first example, the Dockerfile is implementing everything 

Reviewer: ```suggestion
written in C.

In the first example, the Dockerfile is implementing everything 
```

in a single stage, living all the objects and spare software in the image. The 

> final example shows how to convert it into a multi-stage container where the 

Reviewer: ```suggestion
second example shows how to convert it into a multi-stage container, where the 
```

> build happens in the first stage, consequently all the objects and spare software 

Reviewer: ```suggestion
build happens in the first stage, and all the objects and spare software 
```

> stays in the first stage. Finally, the second stage copies the binary from the 

Reviewer: ```suggestion
stay in the first stage. Finally, the second stage copies the binary from the 
```

first stage and ends up with a smaller and optimized image.

This guide assumes you have Docker installed on your computer.

Both examples, single and multi-stage, will use the same file structure:

.. prompt:: text

     ├── single
     │   └── Dockerfile
     ├── multi
     │   └── Dockerfile
     ├── helloworld.c
     └── start.sh

Create the file structure:

.. prompt:: bash host:~$, auto

    host:~$ mkdir example
    host:~$ cd example
    host:~$ mkdir single multi

Create the ``helloworld.c`` file:

.. prompt:: bash host:~$, auto

    host:~$ gedit helloworld.c

**helloworld.c**:

.. prompt:: text

    #include <stdio.h>
    int main()
    {
      printf("hello, world!\n");
    }
 
    /* helloworld.c */

Create the ``start.sh`` file:

.. prompt:: bash host:~$, auto

    host:~$ gedit start.sh

**start.sh**:

.. prompt:: text

    #!/bin/sh
    
    while :
    do
      /build/helloworld
      sleep 5
    done

> Add execute permission to the ``start.sh``:

Reviewer: ```suggestion
Add execute permission to ``start.sh``:
```

.. prompt:: bash host:~$, auto

    host:~$ chmod +x start.sh

Single Stage Container
----------------------

Create the ``Dockerfile`` that implements the single-stage container:

.. prompt:: bash host:~$, auto

    host:~$ gedit single/Dockerfile

**single/Dockerfile**:

.. prompt:: text

    FROM debian:bullseye-slim
    RUN  echo "-------Single-Stage--------------"
    
    #Install packages for the build stage
    RUN apt-get update && \
        apt-get install -y --no-install-recommends build-essential && \
        rm -rf /var/lib/apt/lists/*
    
    RUN mkdir -p /build/
    
    COPY ../helloworld.c /build/
    COPY ../start.sh /build/

    WORKDIR /build/
    
    RUN gcc helloworld.c -o helloworld
    
    ENTRYPOINT ["/build/start.sh"]

> The ``Dockerfile`` is very simple, note that it is installing ``build-essential``, 
> coping the files ``helloworld.c`` and ``start.sh`` to the container image, 
> compaling ``helloworld.c`` and seting the entrypoint to start the ``start.sh`` script.

Reviewer: ```suggestion
The ``Dockerfile`` is very simple. It installs ``build-essential``, 
copies the files ``helloworld.c`` and ``start.sh`` to the container image, 
compiles ``helloworld.c`` and sets the entrypoint to start the ``start.sh`` script.
```

Build the example and check the image size:

.. prompt:: bash host:~$, auto

    host:~$ docker build --tag single:1.0 -f single/Dockerfile .
    host:~$ docker image ls

**Example Output**:

.. prompt:: text

     docker image ls
     REPOSITORY                         TAG             IMAGE ID       CREATED          SIZE
     single                             1.0             ba94763b6fe4   25 seconds ago   351MB

Run the image and open a second terminal:

.. prompt:: bash host:~$, auto

    host:~$ docker run -it --rm --name single single:1.0

**Example Output**:

.. prompt:: text

     hello, world!
     hello, world!
     hello, world!

> In the second terminal, inspect the image and note the files and gcc present in the image:

Reviewer: ```suggestion
In the second terminal, inspect the image and note that the spare files and software, like gcc, are present in the image:
```

.. prompt:: bash host:~$, auto

    docker exec -it single gcc

**Example Output**:

.. prompt:: text

     gcc: fatal error: no input files
     compilation terminated.

Multi-Stage Container
---------------------

Create the ``Dockerfile`` that implements the multi-stage container:

.. prompt:: bash host:~$, auto

    host:~$ gedit multi/Dockerfile

**multi/Dockerfile**:

.. prompt:: text

    FROM debian:bullseye-slim AS build
    RUN  echo "-------Multi-Stage--------------"
    
    #Install packages for the build stage
    RUN apt-get update && \
        apt-get install -y --no-install-recommends build-essential && \
        rm -rf /var/lib/apt/lists/*
    
    RUN mkdir -p /build/
    
    COPY helloworld.c /build/
    
    WORKDIR /build/
    
    RUN gcc helloworld.c -o helloworld
    
    RUN  echo "-------Final Stage--------------"
    FROM debian:bullseye-slim AS final-stage
    
    #Install packages for the final stage
    RUN apt-get update && \
        rm -rf /var/lib/apt/lists/*
    
    RUN mkdir -p /build/
    
    COPY --from=build /build/helloworld /build/
    
    WORKDIR /build/
    
    COPY start.sh /build/
    
    ENTRYPOINT ["/build/start.sh"]

> 

Reviewer: Remove this line
```suggestion
```

In this case, the ``Dockerfile`` is divided into two stages: ``build`` and ``final-stage``. 

> The first stage starts with ``AS build`` after specifying the starting image. Next, 

Reviewer: ```suggestion
The first stage starts with ``AS build`` after specifying the starting image (first line of the Dockerfile). Next, 
```

> it is installing ``build-essential`` and compiling the ``helloworld.c``.

Reviewer: ```suggestion
it installs ``build-essential`` and compiles ``helloworld.c``.
```

> The second stage starts with ``AS final-stage`` right after specifying the image used for the second stage. 

Reviewer: ```suggestion
The second stage starts with ``AS final-stage`` right after specifying the image used for the second stage (line 18 of the Dockerfile). 
```

Finally, it ``COPY`` the ``helloworld`` binary from the first stage using the parameter ``--from=build``.

Build the example and check the image size:

.. prompt:: bash host:~$, auto

    host:~$ docker build --tag multi:1.0 -f multi/Dockerfile .
    host:~$ docker image ls

**Example Output**:

.. prompt:: text

     docker image ls
     REPOSITORY                         TAG             IMAGE ID       CREATED          SIZE
     single                             1.0             ba94763b6fe4   25 seconds ago   351MB
     multi                              1.0             bdeac19070ea   50 minutes ago   80.4MB

> Note the difference between the images singe and multi. A simple ``helloworld`` 

Reviewer: ```suggestion
Note the difference between the `single` and `multi` images. A simple ``helloworld`` 
```

with just the ``build-essential`` software shows how useful a multi-stage container could be.

Run the image and open a second terminal:

.. prompt:: bash host:~$, auto

    host:~$ docker run -it --rm --name multi multi:1.0

**Example Output**:

.. prompt:: text

     hello, world!
     hello, world!
     hello, world!

> In the second terminal, inspect the image and note the files and gcc present in the image:

Reviewer: ```suggestion
In the second terminal, inspect the image and note that only the needed files and software are present in the image (``helloworld.c`` and ``gcc`` are not installed in the final stage):
```

.. prompt:: bash host:~$, auto

    docker exec -it multi ls /build

**Example Output**:

.. prompt:: text

     helloworld  start.sh

.. prompt:: bash host:~$, auto

    docker exec -it single gcc

**Example Output**:

.. prompt:: text

     OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "gcc": executable file not found in $PATH: unknown

> Note also that ``helloworld.c`` is not present and ``gcc`` is not installed in the final stage.

Reviewer: ```suggestion
```
